### PR TITLE
(FM-7954) plugin cache purge for stale environments

### PR DESCRIPTION
**Pull request by Thomas-Franklin**
This will use a timer task to check every 24 hours for environments that have not been used in the last 7 days in the plugins folder.

Since the individual environment folders will be updated when a catalog is executed then only environments which have had no agentless catalogs executed in the last 7 days will get purged.

This should help keep excess plugins down, and active environments will be handled by the plugin handler from the Puppet code, i.e. it handles removing and updating of files per request if there are changes.

### CODE REVIEW

File: spec/unit/ace/plugin_cache_spec.rb [hunk 1]
```diff
@@ -16,6 +16,31 @@
     allow(ACE::ForkUtil).to receive(:isolate).and_yield
   end
 
+  describe '#expire' do
```
Comment (willmeek) [inline spec/unit/ace/plugin_cache_spec.rb]: Nice tests!

File: lib/ace/plugin_cache.rb [hunk 2]
```diff
@@ -75,5 +85,23 @@ def sync_core(environment)
...
+    # /opt/puppetlabs/server/data/ace-server/cache/environments
+    # then the directories within this path, which will be
+    # the puppet environments will be removed if they have
+    # not been modified in the last 7 days
+    # when the purge runs (every 24 hours)
+    def expire(purge_ttl)
+      expired_time = Time.now - purge_ttl
+      cache_dir_mutex.with_write_lock do
+        Dir.glob(File.join(cache_dir, '*')).select { |f| File.directory?(f) }.each do |dir|
+          if File.mtime(dir) < expired_time
```
Comment (DavidS) [inline lib/ace/plugin_cache.rb]: What real criteria do we want to measure on which files to remove? The current ticket design does not answer that.

Comment (DavidS) [reply lib/ace/plugin_cache.rb]: And if you're saying "modified 7 days or more ago", when do the environment directories get modified?

Comment (Thomas-Franklin) [reply lib/ace/plugin_cache.rb]: For 1), what do you mean? Do you mean in terms of logging that an environment dir has been removed?

For 2), the environment directory `mtime` will be updated when that environment is used for a agentless catalog execution.

Comment (DavidS) [reply lib/ace/plugin_cache.rb]: 1) I was hoping for something like "The environment hasn't been _used_ for more than 7 days"

2) oh, I wasn't aware of https://github.com/puppetlabs/ace/blob/755e51f7cd87424108b2210072e42a7577a949ce/lib/ace/plugin_cache.rb#L59

**Pull request merged**